### ISSUE TITLE
Breaking change in 5.15.5?

### ISSUE THREAD

**Issue by austinbutler** [mobxjs/mobx#2423]
**Intended outcome:**

Upgrade from 5.15.4 to 5.15.5, app still runs.

**Actual outcome:**

App runs with 5.15.4, but crashes on startup with 5.15.5. Seems many things that were previously working are now `undefined`.

Maybe side effect of https://github.com/mobxjs/mobx/issues/2386?

**How to reproduce the issue:**

If necessary I can try to put this together later. But given the extent of the breakage I assume it's not just me at the moment.

**Versions**

```
"mobx": "5.15.5",
"mobx-react-lite": "^2.0.7",
"mobx-state-tree": "^3.17.2",
"typescript": "^3.9.7"
```

**Comment by MoonBall** [comment]
@austinbutler Please give a example code. So I can try to find out the cause.

**Comment by MoonBall** [comment]
The https://github.com/mobxjs/mobx/issues/2386 has two modifications about `extendObservable`.
1. doesn't include properties of object's prototype.
2. include non-enumerable properties of object.

**Comment by mweststrate** [comment]
@austinbutler, please do provide a reproduction. Without this issue is inactionable.

**Comment by SDSLeon** [comment]
mobs-state-tree also affected.

`
export const GroupSetting = types.model('GroupSetting', {
  id: types.string,
  descript: types.string,
  name: types.enumeration([...SettingNamesArray]),
  value: types.string,
});

settigns: types.array(GroupSetting);

get serverTimezone() {
      const setting = self.settigns.find(setting => setting.name === 'TimeZone');

      console.log(setting.value);
    },
`

returns wrapped finded value

so I need to do setting.value.id to actually get id

**Comment by austinbutler** [comment]
Sorry about that, seems this issue it wasn't as broadly applicable as I'd assumed (which is good!).

I've put together a quick approximation of how it happens in my app. If you toggle between `5.15.4` and `5.15.5` in `package.json` you can see on .4 it loads fine, on .5 it crashes.

https://codesandbox.io/s/empty-dawn-76v4f

**Comment by mweststrate** [comment]
@urugator would you mind taking a look at this one? From the console logs in https://codesandbox.io/s/hidden-snowflake-2clmz?file=/src/Models.ts, at first sight it looks like the dehancer is not applied with for the `.map` of observable array, (it seems to be applied to the filter), so I guess the result from `filter` is not dehanced in #2348?

**Comment by mweststrate** [comment]
So this probably affects mobx-state-tree users only, as that is afaik the only thing using dehancers.

**Comment by urugator** [comment]
@mweststrate `filter` returns new non-observable array, so the `map` isn't called on observable array ... or?

**Comment by mweststrate** [comment]
Eh yeah I guess 🙂. Anyway it seems that map receives a still boxed value
in that example somehow.

Op do 6 aug. 2020 22:28 schreef urugator <notifications@github.com>:

**Comment by urugator** [comment]
How do I create array with enhancer/dehancer?

**Comment by urugator** [comment]
Nevermind I suppose ... based on MST sources, this seems to be the way:
```javascript
const array = observable([{ value: 1 }, { value: 2 }, { value: 3 }]);
mobx._getAdministration(array).dehancer = ({ value }) => value;
```

**Comment by mweststrate** [comment]
https://github.com/mobxjs/mobx-state-tree/blob/4407fce1d572ad54ac121d6887ea652ef8a18b98/packages/mobx-state-tree/src/types/complex-types/array.ts#L129

**Comment by urugator** [comment]
Will PR tomorrow

**Comment by bobinrinder** [comment]
I assume this is the same issue, for us `array.find()` breaks in MST.

Like after `const result = self.items.find((item) => item.id === dataItem.id);`
we can no longer call `result.someAction()`.
Worked fine in older versions.

**Comment by barbalex** [comment]
We have same issue with array.filter

**Comment by urugator** [comment]
Yes, all of these should be fixed in the PR #2426

Until it's released, either downgrade to 5.15.4 / 4.15.4 or a workaround:
```
array.map(thing => thing).find // .filter/.whatever
```

**Comment by mweststrate** [comment]
Fix has been released as 5.15.6 / 4.15.6. Thanks @urugator!
